### PR TITLE
use has_fma intrinsic

### PR DESCRIPTION
**Pull request by oscardssmith**
This removes all bad `fma` checks in `Base` and replaces them with fma multiversioned ones. It also adds `Base.Math.two_mul` which gives the fastest compensated multiplication available regardless of whether the user has `fma` hardware. This is what should be used when possible for compensated multiplication in the future.

### CODE REVIEW

File: base/math.jl [hunk 1]
```diff
@@ -41,6 +41,24 @@ end
...
+    return Base.twomul(x,y)
+end
+
+@inline function two_mul(x::T, y::T) where T<: Union{Float16, Float32}
+    if Core.Intrinsics.have_fma(T)
+        xy = x*y
+        return xy, fma(x, y, -xy)
+    end
+    xy = widen(x)*y
+    Txy = T(xy)
```
Comment (vtjnash) [inline base/math.jl:58]: would this be short a rounding bit to handle the halfway case without double-rounding it?

**Comment by oscardssmith** [comment]
@vtjnash is this good to merge?

**Comment by vtjnash** [comment]
yep, I trust you

**Comment by vtjnash** [comment]
(and fixed the commit message by using your text from the issue above)

**Comment by oscardssmith** [comment]
> yep, I trust you

That's definitely a mistake...

**Pull request merged**